projects/core/src/main/resources/data/computercraft/lua/rom/modules/main/cc/audio/wav.lua

```diff
@@ -0,0 +1,224 @@
+-- audio.length is the length in samples; divide by sample rate to get seconds
+print("Length:", audio.length / audio.sampleRate, "s")
+
+for chunk in audio.read, 131072 do
+  while not speaker.playAudio(chunk) do
+    os.pullEvent("speaker_audio_empty")
+  end
+end
+```
+]]
+
+local expect = require "cc.expect".expect
+local dfpwm = require "cc.audio.dfpwm"
+
+local str_unpack, str_sub, math_floor = string.unpack, string.sub, math.floor
+
+local dfpwmUUID = "3ac1fa38-811d-4361-a40d-ce53ca607cd1" -- UUID for DFPWM in WAV files
+
+local function uuidBytes(uuid) return uuid:gsub("-", ""):gsub("%x%x", function(c) return string.char(tonumber(c, 16)) end) end
+
+local wavExtensible = {
+    dfpwm = uuidBytes(dfpwmUUID),
+    pcm = uuidBytes "01000000-0000-1000-8000-00aa00389b71",
+    msadpcm = uuidBytes "02000000-0000-1000-8000-00aa00389b71",
+    alaw = uuidBytes "06000000-0000-1000-8000-00aa00389b71",
+    ulaw = uuidBytes "07000000-0000-1000-8000-00aa00389b71",
+    adpcm = uuidBytes "11000000-0000-1000-8000-00aa00389b71",
+    pcm_float = uuidBytes "03000000-0000-1000-8000-00aa00389b71",
+}
+
+local wavMetadata = {
+    IPRD = "album",
+    INAM = "title",
+    IART = "artist",
+    IWRI = "author",
+    IMUS = "composer",
+    IPRO = "producer",
+    IPRT = "trackNumber",
+    ITRK = "trackNumber",
+    IFRM = "trackCount",
+    PRT1 = "partNumber",
+    PRT2 = "partCount",
+    TLEN = "length",
+    IRTD = "rating",
+    ICRD = "date",
+    ITCH = "encodedBy",
+    ISFT = "encoder",
+    ISRF = "media",
+    IGNR = "genre",
+    ICMT = "comment",
+    ICOP = "copyright",
+    ILNG = "language",
+}
+
+--[[- Read WAV data into a table.
+
+The returned table contains the following fields:
+- `codec`: A string with information about the codec used in the file (one of `u8`, `s16`, `s24`, `s32`, `f32`, `dfpwm`)
+- `sampleRate`: The sample rate of the audio in Hz. If this is not 48000, the file will need to be resampled to play correctly.
+- `channels`: The number of channels in the file (1 = mono, 2 = stereo).
+- `length`: The number of samples in the file. Divide by sample rate to get seconds.
+- `metadata`: If the WAV file contains `INFO` metadata, this table contains the metadata.
+Known keys are converted to friendly names like `artist`, `album`, and `track`, while unknown keys are kept the same.
+Otherwise, this table is empty.
+- `read(length: number): number[]...`: This is a function that reads the audio data in chunks.
+It takes the number of samples to read, and returns each channel chunk as multiple return values.
+Channel data is in the same format as `speaker.playAudio` takes: 8-bit signed numbers.
+
+@tparam string data The WAV data to read.
+@treturn table The decoded WAV file data table.
+]]
+local function readWAV(data)
+    expect(1, data, "string")
+    local bitDepth, dataType, blockAlign
+    local temp, pos = str_unpack("c4", data)
+    if temp ~= "RIFF" then error("bad argument #1 (not a WAV file)", 2) end
```

Comment:
I realise CC is a little inconsistent here[^1], but I think we should probably follow what Lua does, and return `nil` + the error message instead of erroring here. 

[^1]: For example, `textutils.unserialiseJSON` should probably return `nil, message` rather than erroring.